source/extensions/filters/udp/dns_filter/dns_parser.cc

```diff
@@ -210,20 +392,198 @@ DnsQueryRecordPtr DnsMessageParser::parseDnsQueryRecord(const Buffer::InstancePt
+                                                         query_rec.class_, ttl, std::move(ipaddr));
+  context->answers_.emplace(query_rec.name_, std::move(answer_record));
+}
+
+void DnsMessageParser::setResponseCode(DnsQueryContextPtr& context,
+                                       const uint16_t serialized_queries,
+                                       const uint16_t serialized_answers) {
+  // If the question is malformed, don't change the response
+  if (context->response_code_ == DNS_RESPONSE_CODE_FORMAT_ERROR) {
+    return;
+  }
+  // Check for unsupported request types
+  for (const auto& query : context->queries_) {
+    if (query->type_ != DNS_RECORD_TYPE_A && query->type_ != DNS_RECORD_TYPE_AAAA) {
+      context->response_code_ = DNS_RESPONSE_CODE_NOT_IMPLEMENTED;
+      return;
+    }
+  }
+  // Output validation
+  if (serialized_queries == 0) {
+    context->response_code_ = DNS_RESPONSE_CODE_FORMAT_ERROR;
+    return;
+  }
+  if (serialized_answers == 0) {
+    context->response_code_ = DNS_RESPONSE_CODE_NAME_ERROR;
+    return;
+  }
+  context->response_code_ = DNS_RESPONSE_CODE_NO_ERROR;
+}
+
+void DnsMessageParser::buildResponseBuffer(DnsQueryContextPtr& query_context,
+                                           Buffer::OwnedImpl& buffer) {
+  // Ensure that responses stay below the 512 byte byte limit. If we are to exceed this we must add
+  // DNS extension fields
+  //
+  // Note:  There is Network::MAX_UDP_PACKET_SIZE, which is defined as 1500 bytes. If we support
+  // DNS extensions, which support up to 4096 bytes, we will have to keep this 1500 byte limit in
+  // mind.
+  static constexpr uint64_t MAX_DNS_RESPONSE_SIZE = 512;
+  static constexpr uint64_t MAX_DNS_NAME_SIZE = 255;
+
+  // Each response must have DNS flags, which spans 4 bytes. Account for them immediately so that we
+  // can adjust the number of returned answers to remain under the limit
+  uint64_t total_buffer_size = sizeof(DnsHeaderFlags);
+  uint16_t serialized_answers = 0;
+  uint16_t serialized_queries = 0;
+
+  Buffer::OwnedImpl query_buffer{};
+  Buffer::OwnedImpl answer_buffer{};
+
+  ENVOY_LOG(trace, "Building response for query ID [{}]", query_context->id_);
+
+  for (const auto& query : query_context->queries_) {
+    if (!query->serialize(query_buffer)) {
+      ENVOY_LOG(debug, "Unable to serialize query record for {}", query->name_);
+      continue;
+    }
+
+    // Serialize and account for each query's size. That said, there should be only one query.
+    ++serialized_queries;
+    total_buffer_size += query_buffer.length();
+
+    for (const auto& answer : query_context->answers_) {
+      // Query names are limited to 255 characters. Since we are using ares to decode the encoded
+      // names, we should not end up with a non-conforming name here.
+      //
+      // See Section 2.3.4 of https://tools.ietf.org/html/rfc1035
+      if (query->name_.size() > MAX_DNS_NAME_SIZE) {
```

Comment:
I think I commented on this at some point, but the way that most DNS servers deal with this case AFAIK is to return a random set of hosts that fit into the record limit. I think we should probably do that eventually. Can you make sure there is a relevant TODO about that type of behavior? If that would require any significant structural changes to the code you have here you might want to check that now.

I would also consider stats around things like record overflow.

Reply:
Noted.  Added a TODO for the next PR that includes stats.

I can add the randomizing behavior.  I don't think it will take significant changes.